content/blog/2020-12-19-asmm/index.md

```diff
@@ -0,0 +1,119 @@
+
+Once all of the pointer types are fully annotated, the program can be quickly checked for some of the current limitations. For instance, when you call ```alloc```, we can check that all of the exit blocks that can be reached from this block are domination by the current block. This is to ensure when we come through with the next phase, any call to ```alloc``` is valid and needs to be free at each of the exit blocks dominated by this. We go through one more time and remove all of the region annotations to bring the Bril extension back to standard Bril code.
+
+## Evaluation
+
+Surprisingly, the inference part of this tool is strong enough that all four of the current benchmarks that use the memory extension can be converted to this extension of Bril merely by removing the ```free``` calls. The inference pass is able to annotate the rest in these case, ```eigth-queens.bril, fib.bril. mat-mul.bril, and sieve.bril```. The other benchmarks in the current suite do not use the memory extension but they can still be handed to and returned from this pass with no change. Three out of four of these benchmarks have the exact same total dynamic instruction count between the current benchmark and running this bril extension benchmark through the analysis. ```Sieve.bril``` was slower but this was due to the reconstruction of the program from the control graph being suboptimal and is not a performance issue with the extension itself.
+
+With the exception of the ```mat-mul.bril``` benchmark, these benchmarks are not a very difficult evaluation. It may just be that full flexibility in memory management is hard to keep track of anyways so users are already restricting their use and these benchmarks are representative of some standard use. To compensate for this, I've gone and included the interpreter memory test cases to get fuller coverage of some of the edge cases that could occur. For instance, ```alloc_many.bril``` is a go example of a program that would fail to check in this bril extension.
+
+Alloc_many.bril
+
+```
+@main {
+  v: int = const 1;
+  max: int = const 1000000;
+  count: int = const 0;
+.lbl:
+  count: int = add count v;
+  p: ptr<int> = alloc v;
+  free p;
+  loop: bool = ge count max;
+  br loop .end .lbl;
+.end:
+  print count;
+}
+
+```
+
+Notice how this program loops over the loop body some `max` number of times. Each time, the program allocates and frees some memory. Because the same variable destination is being used each time, this would not check. The allocation would need to be moved to the loop header and either freed immediately or at block `.end`. If there were user defined regions that you could write something like:
+
+Alloc_many.bril.asmm
+
+```
+@main {
+  v: int = const 1;
+  max: int = const 1000000;
+  count: int = const 0;
+.lbl:
+  start_region "loop_body"
```

Comment:
Some discussion of how this might work would be helpful… would these need to be lexically scoped, somehow?

Reply:
`alloc` dominates all of their exit blocks, `start_region` would have to dominate all of the `end_region` instructions. I will add a small blurb.